Use CUDA buffer IDs to validate rcache entries

SeyedMir opened the pull request:
## What
This PR associates a memory attribute to each rcache entry. The attribute enables rcache to use CUDA buffer IDs to validate entries that correspond to CUDA memory allocations.

## Why ?
If UCM hooks fail to install successfully, CUDA memory release functions are not intercepted. Therefore, the corresponding rcache entries are not invalidated. If a new CUDA allocation happens to use the same VA range as the one in rcache, it will lead to an invalid cache hit. This can be avoided by using CUDA buffer IDs as an external mechanism to validate rcache results.

## How ?
Three new callbacks are added to rcache ops, and are used when rcache memory hooks fail to install:
- `mem_reg_ext_validate` used for registration callback
- `mem_dereg_ext_validate` used for deregistration callback 
- `mem_ext_validate` used to validate a cache hit

IB and gdrcopy memory domains will use UCT memory attributes (https://github.com/openucx/ucx/pull/6306) to implement the above callbacks.
- `mem_reg_ext_validate` callback queries memory attribute and adds it to rcache region
- `mem_dereg_ext_validate` callback destroys the memory attribute
- `mem_ext_validate` callback compares the memory attribute of the queried address with the memory attribute of the region returned by rcache.

swx-jenkins3 commented:
Can one of the admins verify this patch?

shamisp commented:
Hi @SeyedMir . Thanks for the patch. What is your affiliation ? thanks !

yosefe commented:
> Hi @SeyedMir . Thanks for the patch. What is your affiliation ? thanks !

NVIDIA

yosefe commented:
ok to test

shamisp commented:
@yosefe - can we ask Nvidia folks to update their GitHub profiles ?

yosefe commented:
> @yosefe - can we ask Nvidia folks to update their GitHub profiles ?

![image](https://user-images.githubusercontent.com/2255631/99415034-14c31480-2900-11eb-9b0f-415b50589d3f.png)

shamisp commented:
@yosefe - it was not there yesterday, I checked.

SeyedMir commented:
That's right. I update it yesterday after Pavel's comment.

SeyedMir commented on test/gtest/ucs/test_rcache.cc:
```diff
@@ -68,7 +69,7 @@ class test_rcache : public ucs::test {
             UCS_PGT_ADDR_ALIGN,
             ucs_get_page_size(),
-            UCM_EVENT_VM_UNMAPPED,
+            UCM_EVENT_VM_UNMAPPED | UCM_EVENT_MEM_TYPE_FREE,
```
Should we use `UCM_LEVEL_MEM_TYPE_FREE` in rcache tests? On one hand we should, because actual use cases of rcache (e.g., IB, gdrcopy) register that event. On the other hand, not using it allows us to test rcache under circumstances similar to when the CUDA runtime is linked statically.
Or maybe we should repeat the tests with and without `UCM_EVENT_MEM_TYPE_FREE`.

SeyedMir commented:
@yosefe do you know why `io_demo: base` and `Read the Docs build` are failing? I cannot figure it out from the CI logs.

yosefe commented:
/azp run

yosefe commented:
bot:pipe:retest

yosefe commented:
will be replaced by bistro hooks for cuda to work with static link

SeyedMir commented:
@bureddy @yosefe can you please review when possible? Thanks

bureddy commented on src/ucm/cuda/cudamem.c:
```diff
@@ -286,22 +293,32 @@ static ucs_status_t ucm_cudamem_install(int events)
-            ucm_warn("failed to install cuda memory hooks on runtime API")
-        }
+        goto out_unlock;
+    }
```
seems this logic is incorrect

(1) is bistro fails with driver funcs,  it is not sufficient with just runtime funcs bistro because the application might be using driver app
(2) with "reloc",  both driver and runtime func needs to be successfully installed to consider ucm_cudamem_installed = 1

bureddy commented on src/ucm/cuda/cudamem.c:
```diff
@@ -211,18 +214,31 @@ static ucm_mmap_hook_mode_t ucm_cuda_hook_mode()
     ucm_cuda_func_t *func;
     ucs_status_t status;
     void *func_ptr;
 
```
`if (hook_mode == UCM_MMAP_HOOK_NONE) { return }`

SeyedMir replied:
`ucm_cudamem_install()` already does that check. Do we need it again here?

bureddy replied:
missed it. then we are good.

SeyedMir commented:
@yosefe @bureddy rcache unit tests will not test buffer id logic because it is disabled by default. How can I make CI run those tests with `UCX_RCACHE_CHECK_MEMATTR` set to `on` and the memory hooks disabled? Should I change `contrib/test_jenkins.sh`?

SeyedMir committed:
UCT/MD: Add a new UCT memory query API

Unlike the existing uct_md_mem_query(), then new API is not specific
to a single MD. Therefore, it does not need a handle to an opened MD.

SeyedMir committed:
UCM/CUDA: Update UCM CUDA hooks

SeyedMir committed:
UCS/RCACHE: Add external validation callbacks to rcache

SeyedMir committed:
UCS/RCACHE: Add rcache external validation config

SeyedMir committed:
TEST/RCACHE: Update rcache tests

SeyedMir committed:
UCT/MD: Add external rcache validation to memory domains

SeyedMir commented:
@yosefe @Akshay-Venkatesh @bureddy I updated this PR to use UCT memory attributes (from https://github.com/openucx/ucx/pull/6306) to validate rcache results when CUDA UCM hooks fail.  
Please let me know your comments.

gleon99 commented:
Hi @SeyedMir - we're cleaning up the UCX PR backlog and plan to close this PR in 7 days. Please let us know if you'd like to keep it open.
